californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -192,195 +211,256 @@ public void onAcknowledgement() {
 
-						exchange.setCurrentResponse(piggybacked);
-						lower().sendResponse(exchange, piggybacked);
+			} else if (!status.addBlock(request.getPayload())) {
 
-						// do not assemble and deliver the request yet
+				sendBlock1ErrorResponse(key, exchange, request, block1, ResponseCode.REQUEST_ENTITY_TOO_LARGE,
+						"body exceeded expected size " + status.getBufferSize());
 
-					} else {
-						LOGGER.finer("This was the last block. Deliver request");
+			} else {
 
-						// Remember block to acknowledge. TODO: We might make this a boolean flag in status.
-						exchange.setBlock1ToAck(block1); 
+				status.setCurrentNum(status.getCurrentNum() + 1);
+				if ( block1.isM() ) {
 
-						// Block2 early negotiation
-						earlyBlock2Negotiation(exchange, request);
+					// do not assemble and deliver the request yet
 
-						// Assemble and deliver
-						Request assembled = new Request(request.getCode());
-						assembled.setSenderIdentity(request.getSenderIdentity());
-						assembleMessage(status, assembled);
+					LOGGER.log(Level.FINE, "acknowledging incoming block1 [num={0}], expecting more blocks to come", block1.getNum());
 
-						exchange.setRequest(assembled);
-						upper().receiveRequest(exchange, assembled);
-					}
+					Response piggybacked = Response.createResponse(request, ResponseCode.CONTINUE);
+					piggybacked.getOptions().setBlock1(block1.getSzx(), true, block1.getNum());
+					piggybacked.setLast(false);
+
+					exchange.setCurrentResponse(piggybacked);
+					lower().sendResponse(exchange, piggybacked);
 
 				} else {
-					Response error = Response.createResponse(request, ResponseCode.REQUEST_ENTITY_INCOMPLETE);
-					error.getOptions().setBlock1(block1.getSzx(), block1.isM(), block1.getNum());
-					error.setPayload("unexpected Content-Format");
 
-					exchange.setCurrentResponse(error);
-					lower().sendResponse(exchange, error);
-					return;
-				}
+					LOGGER.log(Level.FINE, "peer has sent last block1 [num={0}], delivering request to application layer", block1.getNum());
 
-			} else {
-				// ERROR, wrong number, Incomplete
-				LOGGER.log(Level.WARNING,
-						"Wrong block number. Expected {0} but received {1}. Respond with 4.08 (Request Entity Incomplete)",
-						new Object[]{status.getCurrentNum(), block1.getNum()});
-				Response error = Response.createResponse(request, ResponseCode.REQUEST_ENTITY_INCOMPLETE);
-				error.getOptions().setBlock1(block1.getSzx(), block1.isM(), block1.getNum());
-				error.setPayload("Wrong block number");
-				exchange.setCurrentResponse(error);
+					// Remember block to acknowledge. TODO: We might make this a boolean flag in status.
+					exchange.setBlock1ToAck(block1); 
+
+					// Assemble and deliver
+					Request assembled = new Request(request.getCode());
+					assembled.setSenderIdentity(request.getSenderIdentity());
+					status.assembleMessage(assembled);
```

Comment:
For block2 I would agree to use MID/token from the first response, as long as no RST is sent for that response.
If we would be able to RST the response, we need to use the MID from the last.